### ISSUE TITLE
Add lower bounds on dependencies

### ISSUE THREAD

**Issue by rg936672** [gchq/coreax#807]
### What's the issue?

Currently, most of Coreax's dependencies don't have lower bounds specified, so aren't accurate - we clearly don't support very early versions of most of our dependencies. Add minimum versions for each, ideally as loose as possible. This can be tested with uv's `--resolution lowest` strategy. - see #918.

Since automated testing of this is a separate ticket, this ticket can be resolved by just adding naive "latest major version" bounds on the dependencies. (Just doing this puts us in a much better position than the majority of the Python ecosystem!)

**Comment by pc532627** [comment]
Blocked by #509
